### PR TITLE
Jelena's refactor branch

### PR DESCRIPTION
**Pull request by jonathan-taylor**
Comparison with Jelena's branch.

So we can add comments, etc. and track differences

### CODE REVIEW

File: selection/randomized/query.py [hunk 1]
```diff
@@ -1838,3 +1863,117 @@ def _weights(self, candidate):
...
+    weights for a given candidate will look like
+          randomization_density(sample + (candidate, 0, 0) - (reference, 0, 0)) /
+          randomization_density(sample)
+    if the samples are samples of \bar{\beta}. if we have samples of
+    \Delta from our reference, then the weights will look like
+    randomization_density(sample + (candidate, 0, 0))
+    randomization_density(sample + (reference, 0, 0))
+    WE ARE ASSUMING sample is sampled from targeted_sampler.reference
+    """
+
```
Comment (jonathan-taylor) [inline selection/randomized/query.py:1883]: Are we using targeted_sampler just to get an idea of target? 

Or is targeted sampler meant to have computed the cross terms in the covariance as well?

File: selection/randomized/query.py [hunk 2]
```diff
@@ -1838,3 +1863,117 @@ def _weights(self, candidate):
...
+        self.observed = observed.copy() # this is our observed unpenalized estimator
+        nactive = targeted_sampler.observed_target_state.shape[0]
+
+        self._normal_sample = np.random.multivariate_normal(mean=np.zeros(nactive), cov=targeted_sampler.target_cov, size =(sample.shape[0]))
+        print(self._normal_sample.shape)
+        self._sample = np.concatenate((sample, np.tile(self.observed, (sample.shape[0], 1))), axis=1)
+        self._logden = targeted_sampler.log_randomization_density(self._sample)
+        self._delta = np.concatenate((sample, self._normal_sample), axis=1)
+
+
```
Comment (jonathan-taylor) [inline selection/randomized/query.py:1899]: What is candidate supposed to be here?

File: selection/randomized/query.py [hunk 3]
```diff
@@ -1838,3 +1863,117 @@ def _weights(self, candidate):
...
+        upper = bisect(_rootU, grid_min, grid_max, xtol=1.e-5*(grid_max - grid_min))
+        lower = bisect(_rootL, grid_min, grid_max, xtol=1.e-5*(grid_max - grid_min))
+
+        return lower + projected_observed, upper + projected_observed
+
+    # Private methods
+
+    def _weights(self, linear_func, candidate):
+
```
Comment (jonathan-taylor) [inline selection/randomized/query.py:1960]: Shouldn't we see some cross terms in the covariance somewhere?

File: selection/randomized/query.py (hunk 2 above)
Comment (jelena-markovic) [reply selection/randomized/query.py:1899]: Candidate is a parameter for which we evaluate the pivot.

File: selection/randomized/query.py (hunk 1 above)
Comment (jelena-markovic) [reply selection/randomized/query.py:1883]: targeted_sampler is used getting target_cov, but its use was mostly for using this method target_sampler.log_randomization_density() to evaluate the weights.

File: selection/randomized/query.py (hunk 3 above)
Comment (jelena-markovic) [reply selection/randomized/query.py:1960]: We see them through target_sampler.log_randomization_density(), which knows about the decomposition. Then we can use the sample consisting of optimization variables and the target only as arguments of target_sampler.log_randomization_density().

File: selection/randomized/query.py (hunk 2 above)
Comment (jonathan-taylor) [reply selection/randomized/query.py:1899]: OK

File: selection/randomized/query.py (hunk 1 above)
Comment (jonathan-taylor) [reply selection/randomized/query.py:1883]: Is targeted_sampler an instance of optimization_sampler?

File: selection/randomized/query.py (hunk 3 above)
Comment (jonathan-taylor) [reply selection/randomized/query.py:1960]: I'll have to go through the calculations, but I would have expected residual matrix to involve cross terms of the covariance...

Comment (jelena-markovic) [reply selection/randomized/query.py:1960]: This class is just decomposing the target in terms of target.dot(keep) which 1D direction of interest. The score has been decomposed in terms of the whole target vector $\bar{\beta}_E$ earlier in target_sampler.setup_target(). So the cross-terms of the covariance are used before we call this opt_weighted_intervals class.

File: selection/randomized/query.py (hunk 1 above)
Comment (jelena-markovic) [reply selection/randomized/query.py:1883]: Yes. It gets called in the confidence interval method of optimization_sampler: https://github.com/jelena-markovic/Python-software/blob/refactor/selection/randomized/query.py#L1308

File: selection/randomized/query.py [hunk 4]
```diff
@@ -1838,3 +1863,117 @@ def _weights(self, candidate):
...
+        candidate_sample = self._sample.copy()
+
+        _norm = np.linalg.norm(linear_func)
+        projection_matrix = np.true_divide(np.dot(linear_func, linear_func.T), _norm**2)
+        residual_matrix = np.identity(linear_func.shape[0])-projection_matrix
+        candidate_sample[:, self.targeted_sampler.target_slice] = \
```
Comment (jonathan-taylor) [inline selection/randomized/query.py:1966]: I don't understand why you are using this slice of the sample? If ndraw=1000,p=200 and E=30, then S has shape (1000, 200) and its columns have nothing really to do with E. No?

Comment (jelena-markovic) [reply selection/randomized/query.py:1966]: True for S but I defined self._sample to be S concatenated with the observed target and candidate vector is of the same shape as self._sample.

**Pull request merged**